tests/robustness/coverage/key_pattern_test.go

```diff
@@ -0,0 +1,189 @@
+	}
+	if len(s) == 1 && slash == "" {
+		// Get on a non-namespaced resource.
+		return "/registry/" + resource + "/{name}", true
+	}
+	if len(s) == 1 {
+		// Listing all resources in a namespace.
+		return "/registry/" + resource + "/{namespace}" + slash, true
+	}
+	if len(s) == 2 && slash == "" {
+		// Get on a namespaced resource.
+		return "/registry/" + resource + "/{namespace}/{name}", true
+	}
+
+	return k, false
+}
+
+func resourceOrCRD(s string) (string, int) {
+	if strings.Contains(s, ".") {
+		// CRD
+		return "{api-group}/{resource}", 2
+	}
+	switch s {
+	case "services":
+		// services have subresources specs and endpoints.
+		return "{resource}", 2
+	case "masterleases":
+		// masterleases are a special case as they are not a regular (namespaced) resource.
+		return "masterleases", 1
+	case "events":
+		// events are a special case as they have watch cache disabled by default.
+		return "events", 1
+	}
+	return "{resource}", 1
+}
+
+func TestPatternValid(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		key  string
+		want string
+	}{
+		{
+			name: "special_root_key",
+			key:  "compact_rev_key",
+			want: "compact_rev_key",
+		},
+		{
+			name: "special_healthcheck",
+			key:  "/registry/health",
+			want: "/registry/health",
+		},
+		{
+			name: "core_namespace_get",
+			key:  "/registry/controllerrevisions/kube-system/kube-proxy-5884f58745",
+			want: "/registry/{resource}/{namespace}/{name}",
+		},
+		{
+			name: "core_namespace_list",
+			key:  "/registry/controllerrevisions/kube-system/",
+			want: "/registry/{resource}/{namespace}/",
+		},
+		{
+			name: "core_subresource_namespace_get",
+			key:  "/registry/services/specs/default/kubernetes",
+			want: "/registry/{resource}/{namespace}/{name}",
+		},
+		{
+			name: "core_subresource_rev",
+			key:  "/registry/services/endpoints",
+			want: "/registry/{resource}",
+		},
+		{
+			name: "core_subresource_namespace_list",
+			key:  "/registry/services/specs/default/",
+			want: "/registry/{resource}/{namespace}/",
+		},
+		{
+			name: "core_no_namespace_get",
+			key:  "/registry/priorityclasses/system-node-critical",
+			want: "/registry/{resource}/{name}",
+		},
+		{
+			name: "core_no_namespace_rev",
+			key:  "/registry/priorityclasses",
+			want: "/registry/{resource}",
+		},
+		{
+			name: "core_no_namespace_list",
+			key:  "/registry/priorityclasses/",
+			want: "/registry/{resource}/",
+		},
+		{
+			name: "crd_namespace_get",
+			key:  "/registry/cilium.io/ciliumendpoints/kube-system/kube-dns",
+			want: "/registry/{api-group}/{resource}/{namespace}/{name}",
+		},
+		{
+			name: "crd_namespace_list",
+			key:  "/registry/cilium.io/ciliumendpoints/kube-system/",
+			want: "/registry/{api-group}/{resource}/{namespace}/",
+		},
+		{
+			name: "crd_no_namespace_get",
+			key:  "/registry/apiregistration.k8s.io/apiservices/v1alpha1.storage.k8s.io",
+			want: "/registry/{api-group}/{resource}/{name}",
+		},
+		{
+			name: "crd_no_namespace_rev",
```

Comment:
Maybe let's describe what it included and not what's is missing. Like "crd_apigroup_resource"

Reply:
Removed test case names in favor of descriptive values of keys